internal/sysfs/sock.go

```diff
@@ -0,0 +1,66 @@
+package sysfs
+
+import (
+	"net"
+	"syscall"
+
+	"github.com/tetratelabs/wazero/internal/fsapi"
+	"github.com/tetratelabs/wazero/internal/platform"
+)
+
+// SockAccept exposes syscall.Accept on POSIX systems.
+func SockAccept(f fsapi.File) (net.Conn, syscall.Errno) {
+	conn, err := f.(*listenerFile).ln.Accept()
+	if err != nil {
+		return nil, platform.UnwrapOSError(err)
+	}
+	return conn, 0
+}
+
+// SockSetNonblock exposes syscall.SetNonblock on POSIX systems.
+func SockSetNonblock(conn net.Conn) syscall.Errno {
+	tcpConn, ok := conn.(*net.TCPConn)
+	if !ok {
+		return syscall.EINVAL // FIXME: better errno?
+	}
+	syscallConn, err := tcpConn.SyscallConn()
+	if err != nil {
+		return platform.UnwrapOSError(err)
+	}
+	var actualErr error
+	// Control does not allow to return an error, but it is blocking;
+	// so it is ok to modify the external environment and setting
+	// `err` directly.
+	err = syscallConn.Control(func(fd uintptr) {
+		actualErr = setNonblock(fd, true)
+	})
+	if actualErr != nil {
+		return platform.UnwrapOSError(actualErr)
+	}
+	if err != nil {
+		return platform.UnwrapOSError(err)
+	}
+	return 0
+}
+
+// SockShutdown exposes syscall.Shutdown on POSIX systems.
+func SockShutdown(f fsapi.File, how int) syscall.Errno {
+	// FIXME: can userland shutdown listeners?
```

Comment:
Can listening sockets be shutdown? They can't be read or written to so it's a bit unclear what that would do, the man page seems to indicate that they can't?
```
DESCRIPTION
     The shutdown() call causes all or part of a full-duplex connection on the socket associated with socket to be shut down.
```

```
     [ENOTCONN]         The specified socket is not connected.
```

Reply:
My guess is we have to see what happens. For example, wasmtime just pass directly to the underlying socket rather than guess if something should work or not. https://github.com/bytecodealliance/wasmtime/pull/4776/files#diff-ce08e37280f7832c5b5b018feed315bb2acf9eca12941698adb462b141aa590bR299

Reply:
similar in wasmer https://github.com/wasmerio/wasmer/blob/3adc37381d7a32aa25c4a27740e671ee50ba9fec/lib/wasi/src/syscalls/wasix/sock_shutdown.rs#L25

here's wasmedge https://github.com/WasmEdge/WasmEdge/blob/bc3c5641547fb179bc563f34abcb2b975a80b599/lib/host/wasi/inode-macos.cpp#L1260